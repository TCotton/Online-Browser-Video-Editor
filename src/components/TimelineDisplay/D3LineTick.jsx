import React, { useEffect, useRef } from 'react'
import PropTypes from 'prop-types'
import * as d3 from 'd3'

export const D3LineTick = props => {
  const { data, time, width, height } = props

  const ref = useRef()

  useEffect(() => {
    if (
      data &&
      Array.isArray(data.min_array) &&
      Array.isArray(data.max_array)
    ) {
      const svgElement = d3.select(ref.current)

      const min = data.min_array
      const max = data.max_array

      const parseTime = d3.timeFormat('%M:%S')

      const result = time / data.length
      const perSecond = 1 / result

      const xScale = d3
        .scaleLinear()
        .range([0, width])
        .domain([0, data.length])

      const yScale = d3
        .scaleLinear()
        .range([height, 0])
        .domain([d3.min(min), d3.max(max)])

      const ab = d3
        .axisTop(xScale)
        .ticks(20)
        .tickFormat(d => {
          if (perSecond) {
            return parseTime((d / perSecond) * 1000)
          }
          return null
        })

      const d3Line = d3
        .line()
        .x((d, i) => {
          return xScale(i)
        })
        .y(d => {
          return yScale(d)
        })

      svgElement
        .select('.axis-top')
        .call(ab)
        .selectAll('text')
        .attr('transform', 'translate(10,30) rotate(45)')
        .style('text-anchor', 'end')
        .style('font-size', '1.75ex')
        .style('font-family', 'monospace')
        .style('fill', '#ffffff')

      const svg = svgElement
        .select('.waveform')
        .selectAll('path')
        .data([max])

      svg
        .join('path')
        .attr('class', 'svg-line')
        .attr('fill', '#ffffff')
        .attr('d', d3Line)
    }
  }, [data, time])

  return (
    <div className='timeline-ticks'>
      <svg
        viewBox={`0 0 ${width} ${height}`}
        preserveAspectRatio='xMidYMid slice'
        data-testid='svg'
        ref={ref}
      >
        <g className='axis-top' transform='translate(0,0)' data-testid='axis' />
        <g className='waveform' data-testid='waveform' />
      </svg>
    </div>
  )
}

D3LineTick.propTypes = {
  height: PropTypes.number.isRequired,
  width: PropTypes.number.isRequired,
  time: PropTypes.number.isRequired,
  data: PropTypes.shape({
    min_array: PropTypes.arrayOf(PropTypes.number).isRequired,
    max_array: PropTypes.arrayOf(PropTypes.number).isRequired,
    length: PropTypes.number.isRequired
  }).isRequired
}
