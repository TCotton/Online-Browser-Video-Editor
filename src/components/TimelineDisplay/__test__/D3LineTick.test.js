import React from "react";
import {render} from "@testing-library/react";
import {D3LineTick} from "../D3LineTick";

describe('D3LineTick', () => {
    const props = {
        height: 999,
        width: 999,
        time: 12,
        data: {
            min_array: [1],
            max_array: [2],
            length: 123
        }
    }

    it('should render successfully', () => {
        const {container} = render(<D3LineTick {...props}/>);
        expect(container.querySelector("[data-testid=\"svg\"]")).not.toBeNull();
        expect(container.querySelector("[data-testid=\"axis\"]")).not.toBeNull();
        expect(container.querySelector("[data-testid=\"waveform\"]")).not.toBeNull();
    });

    it('should match snapshot', () => {
        const {baseElement} = render(<D3LineTick {...props}/>);
        expect(baseElement).toMatchSnapshot();
    });

});

