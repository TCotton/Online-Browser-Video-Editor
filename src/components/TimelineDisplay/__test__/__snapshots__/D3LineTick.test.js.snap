// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`D3LineTick should match snapshot 1`] = `
<body>
  <div>
    <div
      class="timeline-ticks"
    >
      <svg
        data-testid="svg"
        preserveAspectRatio="xMidYMid slice"
        viewBox="0 0 999 999"
      >
        <g
          class="axis-top"
          data-testid="axis"
          fill="none"
          font-family="sans-serif"
          font-size="10"
          text-anchor="middle"
          transform="translate(0,0)"
        >
          <path
            class="domain"
            d="M0.5,-6V0.5H999.5V-6"
            stroke="currentColor"
          />
          <g
            class="tick"
            opacity="1"
            transform="translate(0.5,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(41.109756097560975,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(81.71951219512195,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(122.32926829268293,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:01
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(162.9390243902439,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:01
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(203.54878048780486,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:02
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(244.15853658536585,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:02
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(284.7682926829268,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:03
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(325.3780487804878,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:03
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(365.9878048780488,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:04
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(406.5975609756097,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:04
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(447.2073170731707,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:05
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(487.8170731707317,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:05
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(528.4268292682926,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:06
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(569.0365853658536,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:06
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(609.6463414634146,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:07
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(650.2560975609756,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:07
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(690.8658536585366,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:08
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(731.4756097560976,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:08
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(772.0853658536586,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:09
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(812.6951219512194,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:09
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(853.3048780487804,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:10
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(893.9146341463414,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:10
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(934.5243902439024,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:11
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(975.1341463414634,0)"
          >
            <line
              stroke="currentColor"
              y2="-6"
            />
            <text
              dy="0em"
              fill="currentColor"
              style="text-anchor: end; font-size: 1.75ex; font-family: monospace; fill: #ffffff;"
              transform="translate(10,30) rotate(45)"
              y="-9"
            >
              00:11
            </text>
          </g>
        </g>
        <g
          class="waveform"
          data-testid="waveform"
        >
          <path
            class="svg-line"
            d="M0,0Z"
            fill="#ffffff"
          />
        </g>
      </svg>
    </div>
  </div>
</body>
`;
